Tuya improvements

trvrnrth opened the pull request:
## Description:
- Reduces heartbeat interval to every 10 seconds per the official specs to cut down chatter
- Adjusts log levels/placement to make operation clearer (less noise in debug; clearer order in verbose)
- Periodic WiFi status updates so the device can better reflect connection state (eg. thermostat display) including API/MQTT connection ("the cloud") status for devices which support it
- Pushes of local time when the time is synched so the device stays up to date
- Tuya climate support to respect an optional MCU reported datapoint that reports active state
- Support for multi-byte bitmasks/bitmaps/fault datapoints
- Simplified methods to set datapoint values that uses the already known datapoint info to greatly ease use in lambdas and allow for easy setting of enums regardless of length: `id(tuya_id).set_datapoint_value(x, y);`

I plan to add support for raw datapoints as well in order to support setting and reading thermostat schedules but will do so in a separate PR when I find the time.

**Related issue (if applicable):** n/a

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1065

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

jesserockz commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -316,64 +301,156 @@ void Tuya::handle_datapoint_(const uint8_t *buffer, size_t len) {
+  this->wifi_status_ = wifi_status;
+  TuyaCommand wifi_command = TuyaCommand{.cmd = TuyaCommandType::WIFI_STATE};
+  if (wifi_status == WL_CONNECTED) {
+    if (this->protocol_version_ == 0x03) {
+      // Notify that we are connected to "the cloud"
+      wifi_command.payload = std::vector<uint8_t>{0x04};
```
Could this be if there is an active `api` or `mqtt` connection?

trvrnrth replied:
That would be nicer if it's easy enough to query. Although I confess I've no idea if it brings any real world improvement to any devices. I've not looked into it but if it's simple enough why not!

trvrnrth replied:
Took the approach from the status component and added.

trvrnrth commented on esphome/components/tuya/climate/tuya_climate.h:
```diff
@@ -40,11 +45,18 @@ class TuyaClimate : public climate::Climate, public Component {
   optional<uint8_t> current_temperature_id_{};
   float current_temperature_multiplier_{1.0f};
   float target_temperature_multiplier_{1.0f};
+  float hysteresis_{1.0f};
```
The idea of bringing this out here was that it would be easy to add a config to set it. My thermostat actually has it available as a (slightly awkward) enum but support to read from that didn't seem sensible as I would imagine any device providing the value as a datapoint will also provide the active state which is more useful.

ic3man5 commented:
I'm currently testing #1404 and was wonder if this will become the actual merged branch for the issue. If so, what needs to happen to make this pull request happen? Tests need to be written?

trvrnrth committed:
Tuya improvements

kroimon commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -344,53 +347,119 @@ void Tuya::send_empty_command_(TuyaCommandType command) {
+void Tuya::set_datapoint_value(uint8_t datapoint_id, uint32_t value) {
+  ESP_LOGD(TAG, "Setting datapoint %u to %u", datapoint_id, value);
+  optional<TuyaDatapoint> datapoint = this->get_datapoint_(datapoint_id);
+  if (!datapoint.has_value()) {
+    ESP_LOGE(TAG, "Attempt to set unknown datapoint %u", datapoint_id);
+  }
+  if (datapoint->value_uint == value) {
+    ESP_LOGV(TAG, "Not sending unchanged value");
+  }
```
Are there returns missing in these cases to abort setting the value?

trvrnrth replied:
@kroimon Indeed there were! Updated now :)

kroimon commented:
Would be great to get this merged.

I think it provides a few good improvements and also allows the tuya switch component to switch enum type datapoints due to the new value setting logic relying on known datapoints and their types.

trvrnrth committed:
Actually don't send unchanged values

kroimon commented on esphome/components/tuya/climate/tuya_climate.cpp:
```diff
@@ -94,30 +102,25 @@ void TuyaClimate::compute_state_() {
-      target_action = climate::CLIMATE_ACTION_IDLE;
-    }
-  } else if (too_hot) {
-    // too hot -> show as cooling if possible, else idle
-    if (this->traits().supports_mode(climate::CLIMATE_MODE_COOL)) {
+    } else if (this->active_state_ == this->active_state_cooling_value_) {
```
`active_state_heating_value_` and `active_state_cooling_value_` are optional and might not have a value assigned here.
You should check for `has_value()` here and also add proper validation to the Python code generation.

trvrnrth replied:
Good idea. I've beefed up the checks here and in the validation. The active heating value does default to 1 as that seemed like an obvious choice (for my use case/device at any rate) but I've left all the checks in place anyway. I was sort of on the fence about having a default for that at all but I figured it would do no harm.

trvrnrth committed:
Beef up climate validation/checks for active state values

trvrnrth committed:
Only allow entering heating/cooling mode if supported

trvrnrth committed:
Lint

trvrnrth committed:
Merge branch 'dev' into tuya-improvements

OttoWinter reviewed (commented):
Thanks! I had a look at this code, but unfortunately posted the comments in the wrong PR. Please see the comments here: #1669. They also apply here.

(for the future, please split such larger code changes into multiple PRs; for example, new features to the climate component should be separate from sending time periodically; otherwise reviewing PRs gets hard)

trvrnrth commented:
> Thanks! I had a look at this code, but unfortunately posted the comments in the wrong PR. Please see the comments here: #1669. They also apply here.
> 
> (for the future, please split such larger code changes into multiple PRs; for example, new features to the climate component should be separate from sending time periodically; otherwise reviewing PRs gets hard)

Yeah; this grew a few extra legs that I wasn't originally planning. I'll address the comments.

kroimon commented on esphome/components/tuya/climate/tuya_climate.cpp:
```diff
@@ -9,55 +9,60 @@ static const char *TAG = "tuya.climate";
   }
   if (this->target_temperature_id_.has_value()) {
     this->parent_->register_listener(*this->target_temperature_id_, [this](TuyaDatapoint datapoint) {
+      ESP_LOGV(TAG, "MCU reported target temperature is: %.1f", this->target_temperature);
       this->target_temperature = datapoint.value_int * this->target_temperature_multiplier_;
```
Swap these lines to get the debug message right.

kroimon commented on esphome/components/tuya/climate/tuya_climate.cpp:
```diff
@@ -9,55 +9,60 @@ static const char *TAG = "tuya.climate";
   }
   if (this->current_temperature_id_.has_value()) {
     this->parent_->register_listener(*this->current_temperature_id_, [this](TuyaDatapoint datapoint) {
+      ESP_LOGV(TAG, "MCU reported current temperature is: %.1f", this->current_temperature);
       this->current_temperature = datapoint.value_int * this->current_temperature_multiplier_;
```
Swap these lines to get the debug message right.

kroimon commented on esphome/components/tuya/climate/tuya_climate.cpp:
```diff
@@ -9,55 +9,60 @@ static const char *TAG = "tuya.climate";
 
 void TuyaClimate::control(const climate::ClimateCall &call) {
   if (call.get_mode().has_value()) {
     this->mode = *call.get_mode();
```
Another thing I just noticed while working with the code:
It is unnecessary to set `this->mode` and `this->this->target_temperature` and call `compute_state_()` and `publish_state()` in `TuyaClimate::control()` as they will be updated in the listeners once the MCU reports back the changed datapoints.

trvrnrth replied:
True. It's probably more accurate to wait until that happens and prevents any possible flip-flopping.

kroimon replied:
Then also replace the assignment of `mode` and `target_temperature` with local const variables.

trvrnrth committed:
Refactor cloud connection check into helper

Also gate local time send behind define.

trvrnrth committed:
Fix logging of MCU reported target/current temp

We now once again correctly log after application of the multiplier.

trvrnrth committed:
Add missing header entry

trvrnrth committed:
Only update state when reported by MCU

We'll always get updates back after our sets so this will more
accurately reflect what the device is actually doing.

trvrnrth committed:
Defer updating climate device state until updates from MCU

trvrnrth committed:
Simplify mode update in control

trvrnrth committed:
Target temp is const in control

trvrnrth committed:
Refactor "cloud" connection helpers

Provide separate helpers for API and MQTT for targeted use and a
consolidating "remote" connection checker.

The consolidated helper will now return true if either type is connected
rather than both to cater for configs which make primary use of MQTT but
still have API configured for the sole purpose of the logger.

trvrnrth committed:
Lint

trvrnrth commented:
As an example of usage here's an [example configuration for an an ET61-W thermostat](https://pastebin.com/qatJjG97) which covers the primary functionality.

trvrnrth committed:
Merge branch 'dev' into tuya-improvements

trvrnrth committed:
Merge branch 'dev' into tuya-improvements

trvrnrth committed:
Fix tuya fan direction settting

trvrnrth committed:
Fix tuya fan direction setting

jesserockz commented on esphome/components/tuya/climate/__init__.py:
```diff
@@ -59,12 +67,37 @@ def validate_temperature_multipliers(value):
+def validate_active_state_values(value):
+    if CONF_ACTIVE_STATE_DATAPOINT not in value:
+        return value
+    if value[CONF_SUPPORTS_HEAT] and CONF_ACTIVE_STATE_HEATING_VALUE not in value:
+        raise cv.Invalid(
+            (
+                f"{CONF_ACTIVE_STATE_HEATING_VALUE} required if using "
+                f"{CONF_ACTIVE_STATE_DATAPOINT} and device supports heating"
+            )
+        )
```
Considering there is a `default=1` for this config, this code will never execute

trvrnrth replied:
I was a little on the fence about that default and didn't have it to start with hence this validation being in place. But I think it's safe enough and a fairly sane default so I'll take this out.

jesserockz replied:
I am happy either way, default with this removed, or no default and this check exists.

jesserockz commented:
Following on from this, I have flashed these changes to my tuya heater and now I cannot control it, because the datapoints are not sent from the heater to the ESP on startup and they are not in `this->datapoints_` which means when it tries to send the datapoint, it does not know the type/length to send.

trvrnrth committed:
Don't init support states as codegen will always set

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

trvrnrth commented on esphome/components/tuya/fan/tuya_fan.cpp:
```diff
@@ -14,6 +14,11 @@ void TuyaFan::setup() {
   if (this->speed_id_.has_value()) {
     this->parent_->register_listener(*this->speed_id_, [this](TuyaDatapoint datapoint) {
+      ESP_LOGV(TAG, "MCU reported speed of: %d", datapoint.value_enum);
+      if (datapoint.value_enum > 0x02) {
+        ESP_LOGE(TAG, "Speed has invalid value %d", datapoint.value_enum);
+        return;
+      }
```
I think the goalposts have moved since I wrote this. I'll check the current restrictions in dev and amend accordingly.

trvrnrth replied:
Yep, bad merge. Corrected now.

trvrnrth committed:
Bail on attempt to set unknown datapoint

trvrnrth committed:
Fix bad merge of fan speed handling

jesserockz commented:
The issue with bailing out on unknown datapoints is that I will not be able to control my heater...

The heater/tuyaMCU only sends the current temperature every 10 seconds, and the other datapoints only if they are changed (either via datapoint settings, or via pressing the buttons on the heater itself). So in an ideal automated world, I would not touch those buttons on the heater, therefore the datapoints would never be sent/recognised by ESPHome...

trvrnrth committed:
Remove validation for heat active state value

We're providing a likely default here so will always have a value.

trvrnrth commented:
> The issue with bailing out on unknown datapoints is that I will not be able to control my heater...
> 
> The heater/tuyaMCU only sends the current temperature every 10 seconds, and the other datapoints only if they are changed (either via datapoint settings, or via pressing the buttons on the heater itself). So in an ideal automated world, I would not touch those buttons on the heater, therefore the datapoints would never be sent/recognised by ESPHome...

That latest commit was just me fixing the behaviour for the current setup. Does your device not report the current datapoints when requested during the init phases in response to the `DATAPOINT_QUERY`? My understanding of the specs is that this should return all current values (and it does on my thermostat device at any rate).

trvrnrth commented:
Is it possible that your device isn't making it through all the init stages and not getting as far as the datapoint query? Perhaps getting stuck at the wifi stage for some reason? I've not read the full conversation on discord but it seems like having a list of all the datapoints and there values to be able to work with later would help with some other issues!

jesserockz commented:
The issue with the setup stage is that I cannot attach to the logs at setup. But let me try this branch again and see what happens.

jesserockz commented:
Ok, stupid me... At the same time I changed to this branch, I updated my yaml to packages and the logger baud_rate was not getting overridden correctly... All working fine now :facepalm:

jesserockz commented:
> I've not read the full conversation on discord but it seems like having a list of all the datapoints and their values to be able to work with later would help with some other issues!

Nah, I have two different tuya based heaters, one uses this component just fine (I wrote the climate for it when I purchased it) but my newer ones (talked about in discord) use a totally different protocol that is not yet on the internet. Its almost completely incompatible with this component and I have duplicated `tuya` and started writing a `tuya-new` so I could try to figure it out.

jesserockz committed:
Merge branch 'dev' into pr/trvrnrth/1491

jesserockz committed:
Update color_temp datapoint

jesserockz commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -9,7 +10,8 @@ static const char *TAG = "tuya";
 
 void Tuya::setup() {
-  this->set_interval("heartbeat", 1000, [this] { this->send_empty_command_(TuyaCommandType::HEARTBEAT); });
+  this->send_empty_command_(TuyaCommandType::HEARTBEAT);
+  this->set_interval("heartbeat", 10000, [this] { this->send_empty_command_(TuyaCommandType::HEARTBEAT); });
```
The scheduler is set up in a way that it executes the first time now-ish
So this results in 2 heartbeats.

```suggestion
  this->set_interval("heartbeat", 10000, [this] { this->send_empty_command_(TuyaCommandType::HEARTBEAT); });
```

trvrnrth replied:
Ah, cool. That makes that a bit neater. I didn't even think to check!

jesserockz committed:
Remove first/duplicate heartbeat

jesserockz committed:
Move time function declaration into #ifdef

kroimon commented on esphome/components/tuya/tuya.cpp:
```diff
@@ -10,7 +10,6 @@ static const char *TAG = "tuya";
 static const int COMMAND_DELAY = 50;
 
 void Tuya::setup() {
-  this->send_empty_command_(TuyaCommandType::HEARTBEAT);
```
Will this delay the whole initialization by 10 seconds?

kroimon replied:
Sorry, didn't see your comment that the first interval will be shorter than what is given as the argument.

jesserockz replied:
The scheduler sets the `last_execution` at now - interval, so its immediately scheduled

jesserockz committed:
Fic formatting

trvrnrth committed:
Remove duplicate check/send of wifi status on init

The first interval will fire immediately so no need to run manually.
